services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java

```diff
@@ -261,54 +263,71 @@ public static class CRLFileLoader extends CRLLoaderImpl {
                 throw new NullPointerException("Context cannot be null");
         }
 
         public Collection<X509CRL> getX509CRLs() throws GeneralSecurityException {
-            CertificateFactory cf = CertificateFactory.getInstance("X.509");
-            Collection<X509CRL> crlColl = null;
+            CrlStorageProvider crlCache = session.getProvider(CrlStorageProvider.class);
+            X509CRL crl = null;
+            if (cRLPath != null) {
+                crl = crlCache.get(cRLPath, this::loadCRL);
+            }
+
+            if (crl == null) {
+                throw new GeneralSecurityException(String.format("Unable to load CRL from \"%s\"", cRLPath));
+            }
+
+            if (crl.getNextUpdate() != null && crl.getNextUpdate().compareTo(new Date(Time.currentTimeMillis())) < 0) {
+                final String message = String.format("CRL from '%s' is not refreshed. Next update is %s.", cRLPath, crl.getNextUpdate());
+                logger.warn(message);
+                if (abortIfNonUpdated) {
+                    throw new GeneralSecurityException(message);
+                }
+            }
 
+            return Collections.singletonList(crl);
+        }
+
+        private X509CRL loadCRL() throws GeneralSecurityException {
+            CertificateFactory cf = CertificateFactory.getInstance("X.509");
+            X509CRL crl = null;
             if (cRLPath != null) {
                 if (cRLPath.startsWith("http") || cRLPath.startsWith("https")) {
                     // load CRL using remote URI
                     try {
-                        crlColl = loadFromURI(cf, new URI(cRLPath));
+                        crl = loadFromURI(cf, new URI(cRLPath));
                     } catch (URISyntaxException e) {
                         logger.error(e.getMessage());
                     }
                 } else if (cRLPath.startsWith("ldap")) {
                     // load CRL from LDAP
                     try {
-                        crlColl = loadCRLFromLDAP(cf, new URI(cRLPath));
+                        crl = loadCRLFromLDAP(cf, new URI(cRLPath));
                     } catch(URISyntaxException e) {
                         logger.error(e.getMessage());
                     }
                 } else {
                     // load CRL from file
-                    crlColl = loadCRLFromFile(cf, cRLPath);
+                    crl = loadCRLFromFile(cf, cRLPath);
                 }
             }
-            if (crlColl == null || crlColl.size() == 0) {
-                String message = String.format("Unable to load CRL from \"%s\"", cRLPath);
-                throw new GeneralSecurityException(message);
-            }
-            return crlColl;
+            return crl;
         }
 
-        private Collection<X509CRL> loadFromURI(CertificateFactory cf, URI remoteURI) throws GeneralSecurityException {
+        private X509CRL loadFromURI(CertificateFactory cf, URI remoteURI) throws GeneralSecurityException {
```

Comment:
Looks like `loadFromURI`, `loadCRLFromLDAP`, `loadFromURI`, and `loadFromStream` can be static methods as they do not use any fields.

Reply:
Not all of them, `loadFromUri` uses the session and `loadCRLFromLDAP` uses ldapContext. So I have not changed any of them.